tasks/api/src/main/java/org/projectnessie/nessie/tasks/api/TaskState.java

```diff
@@ -69,27 +70,52 @@ public interface TaskState {
   @JsonInclude(JsonInclude.Include.NON_NULL)
   String message();
 
-  TaskState SUCCESS = ImmutableTaskState.of(TaskStatus.SUCCESS, null, null, null);
+  /**
+   * Represents the deadline when a task will fail even if it gets a re-triable error. Note: passing
+   * the deadline may or may not forcibly cancel a running task.
+   */
+  @Value.Parameter(order = 5)
+  @Nullable
+  @JsonInclude(JsonInclude.Include.NON_NULL)
+  @JsonDeserialize(using = InstantAsLongDeserializer.class)
+  @JsonSerialize(using = InstantAsLongSerializer.class)
+  Instant deadline();
+
+  TaskState SUCCESS = ImmutableTaskState.builder().status(TaskStatus.SUCCESS).build();
 
   static TaskState successState() {
     return SUCCESS;
   }
 
-  static TaskState runningState(@Nonnull Instant retryNotBefore, @Nonnull Instant lostNotBefore) {
-    return ImmutableTaskState.of(TaskStatus.RUNNING, retryNotBefore, lostNotBefore, null);
+  private static Instant deadlineFor(@Nullable TaskState old, Supplier<Instant> deadline) {
+    if (old == null || old.deadline() == null) {
+      return deadline.get();
```

Comment:
Do we really need a Supplier for this one?

Reply:
removed this from current PR... will follow-up separately.